src/Features/CSharp/Portable/ImplementInterface/AbstractChangeImplementionCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,267 @@
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.ImplementInterface
+{
+    // A mapping from a concrete member to the interface members it implements.
+    using MemberImplementationMap = OrderedMultiDictionary<ISymbol, ISymbol>;
+
+    internal abstract class AbstractChangeImplementionCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        private static readonly SymbolDisplayFormat NameAndTypeParametersFormat =
+            new SymbolDisplayFormat(
+                globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.Omitted,
+                typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameOnly,
+                genericsOptions: SymbolDisplayGenericsOptions.IncludeTypeParameters,
+                miscellaneousOptions: SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
+
+        protected abstract string Implement_0 { get; }
+        protected abstract string Implement_all_interfaces { get; }
+        protected abstract string Implement { get; }
+
+        protected abstract bool CheckExplicitName(ExplicitInterfaceSpecifierSyntax? explicitName);
+        protected abstract bool CheckMember(ISymbol member);
+        protected abstract SyntaxNode ChangeImplementation(SyntaxGenerator generator, SyntaxNode currentDecl, ISymbol interfaceMember);
+        protected abstract Task UpdateReferencesAsync(Project project, Dictionary<Document, SyntaxEditor> documentToEditor, ISymbol implMember, INamedTypeSymbol containingType, CancellationToken cancellationToken);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (container, explicitName, name) = await GetContainerAsync(context).ConfigureAwait(false);
+            if (container == null)
+                return;
+
+            if (!CheckExplicitName(explicitName))
+                return;
+
+            var (document, _, cancellationToken) = context;
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var member = semanticModel.GetDeclaredSymbol(container, cancellationToken) ??
+                throw new InvalidOperationException();
+
+            if (!CheckMember(member))
+                return;
+
+            var project = document.Project;
+
+            var directlyImplementedMembers = new MemberImplementationMap();
+            directlyImplementedMembers.AddRange(member, member.ExplicitOrImplicitInterfaceImplementations());
+
+            var codeAction = new MyCodeAction(
+                string.Format(Implement_0, member.ExplicitOrImplicitInterfaceImplementations().First().Name),
```

Comment:
Isn't `member.ExplicitOrImplicitInterfaceImplementations().First().Name` just the same as member.Name? in this case? It otherwise seems funny that we're passing in the MultiDictionary but for some reason we're explicitly picking out just the first one for the name.

(Doubly odd that if that is different, why it's not just using the value you computed a few lines earlier...)

Reply:
> Isn't member.ExplicitOrImplicitInterfaceImplementations().First().Name just the same as member.Name? in this case? 

It is *not* :)   `member.Name` can be something like `"System.Collections.Generic.IEnumerable<int>.GetEnumerator"`.  i.e. the compiler puts the fully qualified name of the interface impl in an explicit impl.

Reply:
By design.

Reply:
by design.

Reply:
Then I guess I'm still confused. So if it matters for the explicit -> implicit case, the .First() is equivalent to a .Single(), because the map will always only have single value for the member in that case?

(this really needs a comment it feels like...)

Reply:
> Then I guess I'm still confused. So if it matters for the explicit -> implicit case, the .First() is equivalent to a .Single(), because the map will always only have single value for the member in that case?

I *guess* that may be so.  I *imagine* that in C# it wouldn't be possible at all to have a method with multiple explicit-impls, but i wouldn't put it past the compiler to potentially do it (maybe in error cases).

That said, i'm happy to comment the code to make it clear :)

Reply:
Comment would be great! My primary concern is .First() implies there can be two, but it looks like we'd then handle that wrong, at least in that the code fix description wouldn't say what was going on.=